packages/block-library/src/table-of-contents/edit.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Internal dependencies
+ */
+import * as Utils from './utils';
+import ListLevel from './ListLevel';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { subscribe } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import {
+	IconButton,
+	Toolbar,
+	PanelBody,
+	ToggleControl,
+} from '@wordpress/components';
+import { BlockControls, InspectorControls } from '@wordpress/editor';
+
+class TOCEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			wpDataUnsubscribe: null,
+		};
+
+		this.toggleAttribute = this.toggleAttribute.bind( this );
+		this.refresh = this.refresh.bind( this );
+	}
+
+	toggleAttribute( propName ) {
+		const value = this.props.attributes[ propName ];
+		const { setAttributes } = this.props;
+
+		setAttributes( { [ propName ]: ! value } );
+	}
+
+	refresh() {
+		const { setAttributes } = this.props;
+		const headings = Utils.getPageHeadings();
+		setAttributes( { headings } );
+	}
+
+	componentDidMount() {
+		const { attributes, setAttributes } = this.props;
+		const headings = attributes.headings || [];
+		const wpDataUnsubscribe = subscribe( () => {
+			const pageHeadings = Utils.getPageHeadings();
+			this.setState( { pageHeadings } );
+		} );
+
+		setAttributes( { headings } );
+		this.setState( { wpDataUnsubscribe } );
+	}
+
+	componentWillUnmount() {
+		this.state.wpDataUnsubscribe();
+	}
+
+	componentDidUpdate( prevProps, prevState ) {
```

Comment:
This is no longer updating state so (I think?) doesn't trigger a re-render. It might still be possible to merge into `subscribe` though - what do you think?